### ISSUE TITLE
PATH gets mangled when using direnv from git-bash on Windows

### ISSUE THREAD

**Issue by irontoby** [direnv/direnv#253]
Hello,

Installing the [official `git` release for Windows](https://git-scm.com/downloads) will also install the `git-bash.exe` shell. Rather than an emulation layer, I believe this is `bash` plus lots of the standard Linux tools which have been compiled directly for Windows (it seems to be based off the [Mingw-w64 project](http://mingw-w64.org/doku.php)).

Within this shell, it mounts the Windows `C:` drive as `/c` and converts all the Windows paths to their Bash equivalents. For example my `PATH` right now looks something like this:

	/c/Users/tobyj/bin:/mingw64/bin:/usr/local/bin:/usr/bin:/bin:/mingw64/b
	in:/usr/bin:/c/ProgramData/Oracle/Java/javapath:/c/Windows/system32:/c/
	Windows:/c/Windows/System32/Wbem:/c/Program Files/MariaDB 10.1/bin:/c/P
	rogram Files (x86)/Yarn/bin:/usr/bin/vendor_perl:/usr/bin/core_perl

I downloaded the Windows version of `direnv` (direnv.windows-amd64.exe) and put it in a folder in my path. Then I added `eval "$(direnv hook bash)"` to my `.bashrc`. However when I create an `.envrc` and try either `PATH_add node_modules/.bin` or `export PATH=$PWD/node_modules/.bin:$PATH`, it appears that direnv is converting my `$PATH` to use Windows-style paths instead:

	C:\devs\myproj\node_modules\.bin;C:\Users\tobias.johnson\bin;C:\Progr
	am Files\Git\mingw64\bin;

Is there a way to get the Windows .exe to use Bash-style paths and not try to convert them? Or is there a config setting I need to change to get this behavior?

**Comment by zimbatm** [comment]
Hi, thanks for the comprehensive report.

It's a long shot but can you try the latest release? It's now built with go 1.8 which has improved the windows path support.

I produce the windows binaries because it's easy to do so with Go but I must admit of not having tested it.

**Comment by irontoby** [comment]
Nope, no luck. Knowing that it's not really tested/supported on Windows helps though; if I get a chance I'll try to dig in a bit more and figure out whether it's something that can be worked around. Thanks!

**Comment by zimbatm** [comment]
Cool. I'm very interested in having good Windows support, it's just that I am not using it day-to-day. Maybe also have a look at the existing windows issues, it might give you some more clues: https://github.com/direnv/direnv/issues?q=is%3Aopen+is%3Aissue+label%3AWindows

**Comment by markwest1** [comment]
@zimbatm, I am experiencing the same issue and would be glad to help as well. If I can do any testing or coding (I've no production experience writing golang, but love to tinker with it).

**Comment by zimbatm** [comment]
Can you try to run `direnv expand_path /c/Users/tobyj/bin`? If it converts it to the `C:\` format it means that the golang stdlib does something funky.

Basically I think we would have to replace or expand the `path` module in the standard library to understand more of the windows path intricacies. Maybe windows needs multiple path evaluation modes depending on whenever direnv is being run in cmd.exe, bash, PowerShell, ...

**Comment by markwest1** [comment]
![image](https://cloud.githubusercontent.com/assets/8137897/24002732/8372155a-0a2f-11e7-9524-3a56cb89660c.png)
![image](https://cloud.githubusercontent.com/assets/8137897/24002742/888b99e4-0a2f-11e7-9449-2e5b8fa03642.png)

I'll take a look in the direnv code ... should I start in exec.go ? Or maybe upwards in the call stack?

**Comment by zimbatm** [comment]
That's another issue we'll hit with Windows. Because there is a white space in the $PWD, bash interprets it as multiple arguments. Make sure to put double quotes around $PWD, or any variables in general.

**Comment by markwest1** [comment]
I don't think the line in the shell output, `/usr/bin/bash: line 195 : C:/Program: No such file or directory` is caused by the line in the `.envrc` file, `expand_path $PWD` because the same output is generated when the line is changed to `expand_path "$PWD"`:
![image](https://cloud.githubusercontent.com/assets/8137897/24003801/760fa370-0a32-11e7-84f1-6991d6432509.png)
![image](https://cloud.githubusercontent.com/assets/8137897/24003824/86e6a414-0a32-11e7-9b99-bd8f57a309b2.png)
But I agree with your assertion that double quotes will be necessary around bash variable evaluation statements like `$PWD` in `.envrc` files

**Comment by zimbatm** [comment]
It's possible some of the stdlib.sh isn't escaped properly. Can you simplify the `.envrc` to just `echo SUCCESS`?

**Comment by markwest1** [comment]
![image](https://cloud.githubusercontent.com/assets/8137897/24014152/dbf2cde2-0a51-11e7-8288-9a2ff6b7ea67.png)

**Comment by markwest1** [comment]
Nope, sorry.

**Comment by zimbatm** [comment]
@markwest1 @xn can you try the current master? I think I fixed this particular issue.

**Comment by mcandre** [comment]
I get a similar error when using direnv 2.11.3, either 64-bit or 32-bit, in a Git Bash context.

If I use a blank .envrc, or a .envrc with 100% commented out code, my Git Bash shell works fine.

If I try to source rsvm from .envrc, then the resulting Git Bash shell environment can no longer execute commands. `ls`, `which`, and `type` all complain about missing paths, due to how direnv mangles paths when run in a cygwin environment.

**Comment by zimbatm** [comment]
@mcandre I agree, right now direnv is relying on the golang stdlib path package but I think that it's not sufficient for windows support as windows is not a single target.

If someone is willing to implement a path package that works in all these environments it would go a long way to have direnv supported on windows. Right now I suspect that direnv will only work properly inside of the [WSL](https://msdn.microsoft.com/en-gb/commandline/wsl/install_guide).

**Comment by xn** [comment]
BTW, I got it working.

**Comment by xn** [comment]
in git-bash
```
$ which direnv
/c/Users/IEUser/bin/direnv
```

```
$ cat ~/.bashrc
eval "$(direnv hook bash)"
```

**Comment by zimbatm** [comment]
was the solution to remove whitespaces from the path?

**Comment by xn** [comment]
yes

**Comment by markwest1** [comment]
@zimbatm the issue is resolved when using `direnv` in git-bash (MINGW64) on windows. Thank you; sorry it took so long to get back to you.

**Comment by zimbatm** [comment]
no worries. if you still have whitespaces try direnv v2.12.2, it should also fix the issue

**Comment by vagabundmw** [comment]
I'm still having issues with the current version of direnv in Windows Git Bash. For my use case I want to extend the PATH variable, which also contains spaces. As a crude workaround I patched the prompt hook. So instead of calling `eval "$(direnv hook bash)"` I'm using:

    _direnv_hook() {
      local previous_exit_status=$?;
      eval "$(MSYS_NO_PATHCONV=1 "direnv.exe" export bash | sed 's|export PATH=|export _X_DIRENV_PATH=|g')";
      if [ -n "$_X_DIRENV_PATH" ]; then
        _X_DIRENV_PATH=$(cygpath -p "$_X_DIRENV_PATH")
        export "PATH=$_X_DIRENV_PATH"
        unset _X_DIRENV_PATH
      fi
      return $previous_exit_status;
    };


    if ! [[ "$PROMPT_COMMAND" =~ _direnv_hook ]]; then
      PROMPT_COMMAND="_direnv_hook;$PROMPT_COMMAND"
    fi

This hook feeds PATH assignments through `cygpath`, which converts the Windows format in direnv's output to the format Bash expects.

Also, since direnv is a non-MSYS executable, MSYS will translate other environment variables (according to [these rules](http://www.mingw.org/wiki/Posix_path_conversion)) before calling it. This was undesirable for my scenario so I disable it by setting `MSYS_NO_PATHCONV`.

**Comment by sph3rex** [comment]
Additionally in the latest version, even an empty .envrc on win10 with gitbash triggers this error: 

> direnv: error write 3: The handle is invalid.

**Comment by tonyg** [comment]
Here's a similar hack, replacing `eval $(direnv hook bash)` in `.bashrc`:

```shell
export _unmangle_direnv_names='PATH'
_unmangle_direnv_paths() {
    for k in $_unmangle_direnv_names; do
        eval "$k=\"\$(/usr/bin/cygpath -p \"\$$k\")\""
    done
}
eval "$(direnv hook bash | sed -e 's@export bash)@export bash)\
_unmangle_direnv_paths@')"
```

If path-unmangling is required for other variables, they can be added to `_unmangle_direnv_names` in each `.envrc`.

**Comment by igaki** [comment]
> So instead of calling `eval "$(direnv hook bash)"` I'm using:
In my git-bash (PortableGit-2.47.1-64) environment and direnv (2.35.0), _direnv_hook perfectly resolved my PATH troubles. I am deeply grateful for this.

I hope this trick can be added to the documentation.

**Comment by zartc** [comment]
> I hope this trick can be added to the documentation.

I would rather prefer the call to `/usr/bin/cygpath -p` on a cygwing platform to perform POSIX-like path conversion, to be integrated directly into the output of the `direnv hook` so that the user experience in the .bashrc stays the same for everyone.

**Comment by doronbehar** [comment]
@zimbatm I'd like to help solve this issue so that windows users, on any shell will be able to just use `direnv hook $SHELL` with no workarounds. I believe that the problem is that direnv if OS == windows then the PATH needs to be modified for it like in powershell, but in fact it is the shell used in `direnv export` that should dictate that. I just didn't find the place in the code where `$PATH` is handled specially for windows... Could you maybe give a hand?

**Comment by doronbehar** [comment]
OK managed to fix this with Claude-code:

- https://github.com/direnv/direnv/pull/1595

**Comment by tesujimath** [comment]
Thanks to @vagabundmw for the [most excellent work-around for Bash](https://github.com/direnv/direnv/issues/253#issuecomment-466788426).

Claude helped me adapt this for Fish, as follows (add to `config.fish`):
```
# custom direnv integration for MSYS2
# adapted for fish from
# https://github.com/direnv/direnv/issues/253#issuecomment-466788426
function _direnv_hook --on-event fish_prompt
    set -lx MSYS_NO_PATHCONV 1

    set -l script (direnv.exe export fish | string replace -ar -- '(set\s+(?:-\S+\s+)+)PATH\b' '${1}_X_DIRENV_PATH')

    for stmt in $script
        eval $stmt
    end

    if set -q _X_DIRENV_PATH
        set -gx PATH (string split ':' -- (cygpath -p -- (string join ':' -- $_X_DIRENV_PATH)))
        set -e _X_DIRENV_PATH
    end
end
```
